sycl/source/detail/persistent_device_code_cache.cpp

```diff
@@ -557,6 +557,53 @@ void PersistentDeviceCodeCache::putCompiledKernelToDisc(
     updateCacheFileSizeAndTriggerEviction(getRootDir(), TotalSize);
 }
 
+void PersistentDeviceCodeCache::putDeviceCodeIRToDisc(
+    const std::string &Key, const std::vector<char> &IR) {
+
+  repopulateCacheSizeFile(getRootDir());
+
+  // Do not insert any new item if eviction is in progress.
+  // Since evictions are rare, we can afford to spin lock here.
+  const std::string EvictionInProgressFile =
+      getRootDir() + EvictionInProgressFileSuffix;
+  // Stall until the other process finishes eviction.
+  while (OSUtil::isPathPresent(EvictionInProgressFile))
+    continue;
+
+  // Total size of the item that we are writing to the cache.
+  size_t TotalSize = 0;
+
+  std::string DirName = getDeviceCodeIRPath(Key);
+
+  try {
+    OSUtil::makeDir(DirName.c_str());
+    std::string FileName = DirName + "/ir";
+    std::string FullFileName = FileName + ".bin";
+    LockCacheItem Lock{FileName};
+    if (Lock.isOwned()) {
+      writeBinaryDataToFile(FullFileName, IR);
+      PersistentDeviceCodeCache::trace_KernelCompiler(
+          "device code IR has been cached: ", FullFileName);
+
+      TotalSize = getFileSize(FullFileName);
+      saveCurrentTimeInAFile(FileName + CacheEntryAccessTimeSuffix);
+    } else {
+      PersistentDeviceCodeCache::trace_KernelCompiler("cache lock not owned ",
+                                                      FileName);
+    }
+  } catch (std::exception &e) {
+    PersistentDeviceCodeCache::trace_KernelCompiler(
+        std::string("exception encountered making cache: ") + e.what());
+  } catch (...) {
+    PersistentDeviceCodeCache::trace_KernelCompiler(
+        std::string("error outputting cache: ") + std::strerror(errno));
+  }
+
+  // Update the cache size file and trigger cache eviction if needed.
+  if (TotalSize)
+    updateCacheFileSizeAndTriggerEviction(getRootDir(), TotalSize);
```

Comment:
Is it correct that this happens outside of the lock?

Reply:
This is not my design (CC @cperkinsintel @uditagarwal97), but I'd say yes. The lock we held before is only for a specific entry (directory), and the eviction mechanism employs its own lock for the file containing the total cache size.

Reply:
+1. Eviction uses its own lock.